File: src/fable-library/Observable.ts

import { FSharpChoice$2, Choice_tryValueIfChoice1Of2, Choice_tryValueIfChoice2Of2 } from "./Choice.js";
import { value } from "./Option.js";
import { IDisposable } from "./Util.js";

export interface IObserver<T> {
  OnNext: (x: T) => void;
  OnError: (e: any) => void;
  OnCompleted: () => void;
}

export class Observer<T> implements IObserver<T> {
  public OnNext: (x: T) => void;
  public OnError: (e: any) => void;
  public OnCompleted: () => void;

  constructor(onNext: (x: T) => void, onError?: (e: any) => void, onCompleted?: () => void) {
    this.OnNext = onNext;
    this.OnError = onError || ((_e: any) => { return; });
    this.OnCompleted = onCompleted || (() => { return; });
  }
}

export interface IObservable<T> {
  Subscribe: (o: IObserver<T>) => IDisposable;
}

class Observable<T> implements IObservable<T> {
  public Subscribe: (o: IObserver<T>) => IDisposable;

  constructor(subscribe: (o: IObserver<T>) => IDisposable) {
    this.Subscribe = subscribe;
  }
}

export function protect<T>(f: () => T, succeed: (x: T) => void, fail: (e: any) => void) {
  try {
    return succeed(f());
  } catch (e) {
    fail(e);
  }
}

export function add<T>(callback: (x: T) => void, source: IObservable<T>) {
  source.Subscribe(new Observer(callback));
}

export function choose<T, U>(chooser: (x: T) => U, source: IObservable<T>) {
  return new Observable<U>((observer) =>
    source.Subscribe(new Observer<T>((t) =>
      protect(
        () => chooser(t),
        (u) => { if (u != null) { observer.OnNext(value(u)); } },
        observer.OnError),
      observer.OnError, observer.OnCompleted))) as IObservable<U>;
}

export function filter<T>(predicate: (x: T) => boolean, source: IObservable<T>) {
  return choose((x) => predicate(x) ? x : null, source);
}

export function map<T, U>(mapping: (x: T) => U, source: IObservable<T>) {
  return new Observable<U>((observer) =>
    source.Subscribe(new Observer<T>((t) => {
      protect(
        () => mapping(t),
        observer.OnNext,
        observer.OnError);
    }, observer.OnError, observer.OnCompleted))) as IObservable<U>;
}

export function merge<T>(source1: IObservable<T>, source2: IObservable<T>) {
  return new Observable((observer) => {
    let stopped = false;
    let completed1 = false;
    let completed2 = false;
    const h1 = source1.Subscribe(new Observer<T>(
      (v) => { if (!stopped) { observer.OnNext(v); } },
      (e) => {
        if (!stopped) {
          stopped = true;
          observer.OnError(e);
        }
      },
      () => {
        if (!stopped) {
          completed1 = true;
          if (completed2) {
            stopped = true;
            observer.OnCompleted();
          }
        }
      }));
    const h2 = source2.Subscribe(new Observer<T>(
      (v) => { if (!stopped) { observer.OnNext(v); } },
      (e) => {
        if (!stopped) {
          stopped = true;
          observer.OnError(e);
        }
      },
      () => {
        if (!stopped) {
          completed2 = true;
          if (completed1) {
            stopped = true;
            observer.OnCompleted();
          }
        }
      }));
    return {
      Dispose() {
        h1.Dispose();
        h2.Dispose();
      },
    } as IDisposable;
  }) as IObservable<T>;
}

export function pairwise<T>(source: IObservable<T>) {
  return new Observable<[T, T]>((observer) => {
    let last: T;
    return source.Subscribe(new Observer<T>((next) => {
      if (last != null) {
        observer.OnNext([last, next]);
      }
      last = next;
    }, observer.OnError, observer.OnCompleted));
  }) as IObservable<[T, T]>;
}

export function partition<T>(predicate: (x: T) => boolean, source: IObservable<T>) {
  return [filter(predicate, source), filter((x) => !predicate(x), source)];
}

export function scan<U, T>(collector: (u: U, t: T) => U, state: U, source: IObservable<T>) {
  return new Observable<U>((observer) => {
    return source.Subscribe(new Observer<T>((t) => {
      protect(
        () => collector(state, t),
        (u) => { state = u; observer.OnNext(u); },
        observer.OnError);
    }, observer.OnError, observer.OnCompleted));
  }) as IObservable<U>;
}

export function split<T, U1, U2>(splitter: (x: T) => FSharpChoice$2<U1, U2>, source: IObservable<T>) {
  return [
    choose((v) => Choice_tryValueIfChoice1Of2(splitter(v)), source),
    choose((v) => Choice_tryValueIfChoice2Of2(splitter(v)), source)
  ];
}

export function subscribe<T>(callback: (x: T) => void, source: IObservable<T>) {
  return source.Subscribe(new Observer(callback));
}

File: src/fable-library-rust/src/Observable.rs
pub mod Observable_ {
    use crate::Choice_::Choice_2;
    use crate::Native_::{interface_cast, refCell, Func0, Func1, Func2, Lrc, LrcPtr};
    use crate::System::{Exception, IDisposable, IObservable_1, IObserver_1};

    // -----------------------------------------------------------
    // Observer
    // -----------------------------------------------------------

    #[derive(Clone)]
    pub struct Observer<T: Clone + 'static> {
        on_next: Func1<T, ()>,
        on_error: Func1<LrcPtr<Exception>, ()>,
        on_completed: Func0<()>,
    }

    impl<T: Clone + 'static> core::fmt::Debug for Observer<T> {
        fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
            write!(f, "{}", core::any::type_name::<Self>())
        }
    }

    impl<T: Clone + 'static> core::fmt::Display for Observer<T> {
        fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
            write!(f, "{}", core::any::type_name::<Self>())
        }
    }

    impl<T: Clone + 'static> IObserver_1<T> for Observer<T> {
        fn OnNext(&self, arg0: T) {
            (self.on_next)(arg0)
        }
        fn OnError(&self, arg0: LrcPtr<Exception>) {
            (self.on_error)(arg0)
        }
        fn OnCompleted(&self) {
            (self.on_completed)()
        }
    }

    // -----------------------------------------------------------
    // Observable
    // -----------------------------------------------------------

    #[derive(Clone)]
    pub struct Observable<T: Clone + 'static> {
        subscribe: Func1<LrcPtr<dyn IObserver_1<T>>, LrcPtr<dyn IDisposable>>,
    }

    impl<T: Clone + 'static> core::fmt::Debug for Observable<T> {
        fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
            write!(f, "{}", core::any::type_name::<Self>())
        }
    }

    impl<T: Clone + 'static> core::fmt::Display for Observable<T> {
        fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
            write!(f, "{}", core::any::type_name::<Self>())
        }
    }

    impl<T: Clone + 'static> IObservable_1<T> for Observable<T> {
        fn Subscribe(&self, arg0: LrcPtr<dyn IObserver_1<T>>) -> LrcPtr<dyn IDisposable> {
            (self.subscribe)(arg0)
        }
    }

    // -----------------------------------------------------------
    // Disposable
    // -----------------------------------------------------------

    #[derive(Clone)]
    pub struct Disposable {
        dispose: Func0<()>,
    }

    impl core::fmt::Debug for Disposable {
        fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
            write!(f, "{}", core::any::type_name::<Self>())
        }
    }

    impl core::fmt::Display for Disposable {
        fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
            write!(f, "{}", core::any::type_name::<Self>())
        }
    }

    impl IDisposable for Disposable {
        fn Dispose(&self) {
            (self.dispose)()
        }
    }

    // -----------------------------------------------------------
    // Constructors
    // -----------------------------------------------------------

    pub fn mkObserver<T: Clone + 'static>(
        on_next: Func1<T, ()>,
        on_error: Func1<LrcPtr<Exception>, ()>,
        on_completed: Func0<()>,
    ) -> LrcPtr<dyn IObserver_1<T>> {
        interface_cast!(
            LrcPtr::new(Observer {
                on_next,
                on_error,
                on_completed,
            }),
            Lrc<dyn IObserver_1<T>>,
        )
    }

    fn mkObservable<T: Clone + 'static>(
        subscribe: Func1<LrcPtr<dyn IObserver_1<T>>, LrcPtr<dyn IDisposable>>,
    ) -> LrcPtr<dyn IObservable_1<T>> {
        interface_cast!(
            LrcPtr::new(Observable { subscribe }),
            Lrc<dyn IObservable_1<T>>,
        )
    }

    pub fn mkDisposable(dispose: Func0<()>) -> LrcPtr<dyn IDisposable> {
        interface_cast!(LrcPtr::new(Disposable { dispose }), Lrc<dyn IDisposable>,)
    }

    // A no-op error handler (errors are propagated by combinators when present,
    // but the default subscriber ignores them, mirroring the JS/TS reference).
    pub fn noError() -> Func1<LrcPtr<Exception>, ()> {
        Func1::new(|_e: LrcPtr<Exception>| ())
    }

    pub fn noCompleted() -> Func0<()> {
        Func0::new(|| ())
    }

    // Build an error handler that forwards to a downstream observer.
    fn fwdError<T: Clone + 'static>(
        observer: LrcPtr<dyn IObserver_1<T>>,
    ) -> Func1<LrcPtr<Exception>, ()> {
        Func1::new(move |e: LrcPtr<Exception>| observer.OnError(e))
    }

    // Build a completion handler that forwards to a downstream observer.
    fn fwdCompleted<T: Clone + 'static>(observer: LrcPtr<dyn IObserver_1<T>>) -> Func0<()> {
        Func0::new(move || observer.OnCompleted())
    }

    // -----------------------------------------------------------
    // Combinators
    // -----------------------------------------------------------

    pub fn add<T: Clone + 'static>(callback: Func1<T, ()>, source: LrcPtr<dyn IObservable_1<T>>) {
        source.Subscribe(mkObserver(callback, noError(), noCompleted()));
    }

    pub fn subscribe<T: Clone + 'static>(
        callback: Func1<T, ()>,
        source: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<dyn IDisposable> {
        source.Subscribe(mkObserver(callback, noError(), noCompleted()))
    }

    pub fn choose<T: Clone + 'static, U: Clone + 'static>(
        chooser: Func1<T, Option<U>>,
        source: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<dyn IObservable_1<U>> {
        mkObservable(Func1::new(
            move |observer: LrcPtr<dyn IObserver_1<U>>| -> LrcPtr<dyn IDisposable> {
                let chooser = chooser.clone();
                let obs = observer.clone();
                let inner = mkObserver(
                    Func1::new(move |t: T| {
                        if let Some(u) = chooser(t) {
                            obs.OnNext(u);
                        }
                    }),
                    fwdError(observer.clone()),
                    fwdCompleted(observer.clone()),
                );
                source.Subscribe(inner)
            },
        ))
    }

    pub fn filter<T: Clone + 'static>(
        predicate: Func1<T, bool>,
        source: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<dyn IObservable_1<T>> {
        choose(
            Func1::new(move |x: T| if predicate(x.clone()) { Some(x) } else { None }),
            source,
        )
    }

    pub fn map<T: Clone + 'static, U: Clone + 'static>(
        mapping: Func1<T, U>,
        source: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<dyn IObservable_1<U>> {
        mkObservable(Func1::new(
            move |observer: LrcPtr<dyn IObserver_1<U>>| -> LrcPtr<dyn IDisposable> {
                let mapping = mapping.clone();
                let obs = observer.clone();
                let inner = mkObserver(
                    Func1::new(move |t: T| obs.OnNext(mapping(t))),
                    fwdError(observer.clone()),
                    fwdCompleted(observer.clone()),
                );
                source.Subscribe(inner)
            },
        ))
    }

    pub fn merge<T: Clone + 'static>(
        source1: LrcPtr<dyn IObservable_1<T>>,
        source2: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<dyn IObservable_1<T>> {
        mkObservable(Func1::new(
            move |observer: LrcPtr<dyn IObserver_1<T>>| -> LrcPtr<dyn IDisposable> {
                let stopped = refCell(false);
                let completed1 = refCell(false);
                let completed2 = refCell(false);

                let h1 = {
                    let stopped = stopped.clone();
                    let completed1 = completed1.clone();
                    let completed2 = completed2.clone();
                    let obs_n = observer.clone();
                    let obs_e = observer.clone();
                    let obs_c = observer.clone();
                    let se = stopped.clone();
                    let sc = stopped.clone();
                    source1.Subscribe(mkObserver(
                        Func1::new(move |v: T| {
                            if !*stopped.get() {
                                obs_n.OnNext(v);
                            }
                        }),
                        Func1::new(move |e: LrcPtr<Exception>| {
                            if !*se.get() {
                                se.set(true);
                                obs_e.OnError(e);
                            }
                        }),
                        Func0::new(move || {
                            if !*sc.get() {
                                completed1.set(true);
                                if *completed2.get() {
                                    sc.set(true);
                                    obs_c.OnCompleted();
                                }
                            }
                        }),
                    ))
                };

                let h2 = {
                    let stopped = stopped.clone();
                    let completed1 = completed1.clone();
                    let completed2 = completed2.clone();
                    let obs_n = observer.clone();
                    let obs_e = observer.clone();
                    let obs_c = observer.clone();
                    let se = stopped.clone();
                    let sc = stopped.clone();
                    source2.Subscribe(mkObserver(
                        Func1::new(move |v: T| {
                            if !*stopped.get() {
                                obs_n.OnNext(v);
                            }
                        }),
                        Func1::new(move |e: LrcPtr<Exception>| {
                            if !*se.get() {
                                se.set(true);
                                obs_e.OnError(e);
                            }
                        }),
                        Func0::new(move || {
                            if !*sc.get() {
                                completed2.set(true);
                                if *completed1.get() {
                                    sc.set(true);
                                    obs_c.OnCompleted();
                                }
                            }
                        }),
                    ))
                };

                mkDisposable(Func0::new(move || {
                    h1.Dispose();
                    h2.Dispose();
                }))
            },
        ))
    }

    pub fn pairwise<T: Clone + 'static>(
        source: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<dyn IObservable_1<LrcPtr<(T, T)>>> {
        mkObservable(Func1::new(
            move |observer: LrcPtr<dyn IObserver_1<LrcPtr<(T, T)>>>| -> LrcPtr<dyn IDisposable> {
                let last = refCell::<Option<T>>(None);
                let obs = observer.clone();
                let inner = mkObserver(
                    Func1::new(move |next: T| {
                        if let Some(l) = last.get().clone() {
                            obs.OnNext(LrcPtr::new((l, next.clone())));
                        }
                        last.set(Some(next));
                    }),
                    fwdError(observer.clone()),
                    fwdCompleted(observer.clone()),
                );
                source.Subscribe(inner)
            },
        ))
    }

    pub fn partition<T: Clone + 'static>(
        predicate: Func1<T, bool>,
        source: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<(LrcPtr<dyn IObservable_1<T>>, LrcPtr<dyn IObservable_1<T>>)> {
        let pred = predicate.clone();
        let first = filter(predicate, source.clone());
        let second = filter(Func1::new(move |x: T| !pred(x)), source);
        LrcPtr::new((first, second))
    }

    pub fn scan<U: Clone + 'static, T: Clone + 'static>(
        collector: Func2<U, T, U>,
        state: U,
        source: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<dyn IObservable_1<U>> {
        mkObservable(Func1::new(
            move |observer: LrcPtr<dyn IObserver_1<U>>| -> LrcPtr<dyn IDisposable> {
                let collector = collector.clone();
                let st = refCell(state.clone());
                let obs = observer.clone();
                let inner = mkObserver(
                    Func1::new(move |t: T| {
                        let u = collector(st.get().clone(), t);
                        st.set(u.clone());
                        obs.OnNext(u);
                    }),
                    fwdError(observer.clone()),
                    fwdCompleted(observer.clone()),
                );
                source.Subscribe(inner)
            },
        ))
    }

    pub fn split<T: Clone + 'static, U1: Clone + 'static, U2: Clone + 'static>(
        splitter: Func1<T, LrcPtr<Choice_2<U1, U2>>>,
        source: LrcPtr<dyn IObservable_1<T>>,
    ) -> LrcPtr<(LrcPtr<dyn IObservable_1<U1>>, LrcPtr<dyn IObservable_1<U2>>)> {
        let sp = splitter.clone();
        let first = choose(
            Func1::new(move |v: T| match &*sp(v) {
                Choice_2::Choice1Of2(x) => Some(x.clone()),
                _ => None,
            }),
            source.clone(),
        );
        let second = choose(
            Func1::new(move |v: T| match &*splitter(v) {
                Choice_2::Choice2Of2(x) => Some(x.clone()),
                _ => None,
            }),
            source,
        );
        LrcPtr::new((first, second))
    }
}
